[Patch V2] tests/provisioning: restrict feed access and empower test wrapper scripts

amstewart opened the pull request:
It would seem that Github didn't like my [previous PR](https://github.com/ni/nilrt/pull/60), and for some reason would not accept new commits to the merge branch. This PR is a replacement, which includes the version 2 commits.

---

This patchset does two primary things:

1. It removes the base-feed.conf opkg sources prior to installing the newly build grub-gateway and provisioning test scripts. This keeps the tests fixtures from accidently downloading old (already published) feed content, if the locally-built feeds don't have what is necessary. We would prefer for the test to just fail in that case, rather than falsely passing.
2. It adds a wrapper script for running the migration tests, and empowers it - and the extant test_provisioning.sh script - to pass the lower-leverl tcl-expect script the files they need, and to parse stdout for the test results. The goal here is to (a) allow external builders to easily call these same wrapper scripts and (b) to generally move test logic out of the AZDO component. There will be an associated AZDO PR, which interfaces with the new wrapper scripts.

Testing

Both test_ targets build on my machine.

---

## Version 2
* Also remove the /etc/opkg/NI-dist.conf feed source for the same reason. [commit](https://github.com/ni/nilrt/pull/61/commits/271f22e0e747311deacc1bfb87a5e56f6087af68).

Testing
* Rebuilt test_ targets on my dev machine and confirmed that they still work and look good.

@ni/rtos

amstewart committed:
tests/provisioning: remove opkg base-feeds during test

During the provisioning and migration test runs, the test VM uses the
same base-feeds.conf opkg sources as the canonical image. The
locally-built IPK feeds are applied atop those base feeds. Though the
test VM *should* prefer the locally-built IPKs, it is possible for a
malformed VM or test configuration to invert the priority, and a test
run might pass based upon the outdated content in the base-feeds.

To avoid false positives, add a step to the test scripts to remove the
base-feeds.conf file, so that the only valid opkg sources are the
locally-build IPK feeds.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
tests/provisioning: use the whole test output

It's easier to use to use the entire stdout from the start-vm script
when parsing for test success, rather than messing around passing log
paths to tcl expect.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
test/provisioning: add a helper script for the migration test

Move more of the test logic which runs the migration tests from AZDO
components into the nilrt.git source, so that external builders can run
it.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
test/provisioning; remove debug option from test script

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
tests/provisioning: remove default dist/ feed source

The default dist/ feed source in /etc/opkg/NI-dist.conf could pollute
test runs with old test files. To avoid that happening, remove that feed
source prior to installing dist feed packages.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

shruthi-ravi commented on scripts/tests/provisioning/test_provisioning.sh:
```diff
@@ -9,7 +9,29 @@
 #  * evaluates the ptest results.
 
 script_dir="$(dirname "$0")"
-source "$script_dir/start-vm-using-expect.sh"
```
Do you need to add `set -euo pipefail` here?

amstewart replied:
That's a good idea. Will fix.

shruthi-ravi commented on scripts/tests/provisioning/test_grub_gateway_migration.sh:
```diff
@@ -0,0 +1,63 @@
+		--ipk-feed "${dist_feed}" \
+		"${vm_dir}" \
+		test_safemode_provision.expect
+) | tee $log_file
```
So are there going to be two log files for each test now? start-vm-using-expect.sh also [passes](https://github.com/ni/nilrt/blob/nilrt/master/dunfell/scripts/tests/provisioning/start-vm-using-expect.sh#L145) in a log file name to the expect files. I expected an accompanying change to stop doing that, but maybe I've misunderstood the big picture.

amstewart replied:
That's correct. These new wrapper scripts aren't using that functionality anymore. I didn't bother removing it from the `start-` script because it isn't hurting anything and could potentially be useful to someone using that script directly.

shruthi-ravi replied:
Can you share what the log file names all look like if you have a run on your system?

amstewart replied:
For the migration test:
```
migrate_vm_to_safemode.log  # created by start- script; contains the VM session stdout from the tcl expect script of the same name
test_safemode_provision.log  # created by start- script; contains the VM session stdout from the tcl expect script of the same name
test_grub_gateway_migration.log  # created by the wrapper script; contains a cat of the above two, plus the setup/teardown done in the start- script
```

The test_provisioning wrapper creates similar - but there is only one VM session log.

shruthi-ravi replied:
Thanks. IMO the multiple logs and their names are confusing. But we can fix it later if someone complains enough!

amstewart committed:
tests/provisioning: add safety shell options

Add some safety options to the provisioning test wrapper script, to
handle cases where intermediate functions fail, or script arguments
aren't passed.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
## Version 3
* Add shell options to the `test_provisioning.sh` script, for safety.
